File: src/Core.Rust.Algebra/src/gset.rs
//! G-Set — grow-only set CRDT, the bottom rung of the algebra ladder
//! (G-Set ⊂ Bag ⊂ Z-set). Mirrors `src/Core.TypeScript/g-set/g-set.ts` +
//! `src/Core/GSet.fs`.
//!
//! A G-Set is the Z-set restricted to non-negative multiplicity with no
//! retractions: every element is present exactly once, `union` is the only
//! combiner, and `union` is **idempotent, commutative, and associative** — the
//! three CRDT convergence laws (replicas that have seen the same elements, in any
//! order, converge with no coordination — the monotone, CALM-coordination-free
//! slice).
//!
//! This is the comms substrate of the git-native agent-bus (B-0954): an
//! append-only `docs/agent-bus/` folder of ZetaId-keyed messages IS a G-Set —
//! re-observing the same id is `union` and a no-op.
//!
//! Canonical representation: an **ascending-sorted, duplicate-free** `Vec<T>`
//! under `T: Ord`. The canonical order makes equality a plain element-wise
//! comparison and keeps the cross-language golden vectors byte-stable.

use core::cmp::Ordering;
use core::iter::Sum;
use core::ops::Add;

/// A grow-only set: a canonical ascending-sorted, duplicate-free run under
/// `T: Ord`. The invariant is held by every constructor here — never build one by
/// hand from an unsorted `Vec`; use [`GSet::of_iter`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GSet<T> {
    items: Vec<T>,
}

impl<T: Ord + Clone> GSet<T> {
    /// The empty G-Set (the `union` identity).
    #[must_use]
    pub const fn empty() -> Self {
        GSet { items: Vec::new() }
    }

    /// A one-element G-Set.
    #[must_use]
    pub fn singleton(x: T) -> Self {
        GSet { items: vec![x] }
    }

    /// Canonicalize arbitrary input: sort ascending + drop duplicates.
    pub fn of_iter(xs: impl IntoIterator<Item = T>) -> Self {
        let mut items: Vec<T> = xs.into_iter().collect();
        items.sort();
        items.dedup();
        GSet { items }
    }

    /// Membership via binary search on the sorted run. O(log n).
    #[must_use]
    pub fn contains(&self, x: &T) -> bool {
        self.items.binary_search(x).is_ok()
    }

    /// The CRDT merge: the union of two sorted-unique runs, kept sorted-unique.
    /// Idempotent, commutative, associative.
    #[must_use]
    pub fn union(&self, other: &Self) -> Self {
        if self.items.is_empty() {
            return other.clone();
        }
        if other.items.is_empty() {
            return self.clone();
        }
        let (a, b) = (&self.items, &other.items);
        let mut out: Vec<T> = Vec::with_capacity(a.len() + b.len());
        let (mut i, mut j) = (0usize, 0usize);
        while i < a.len() && j < b.len() {
            match a[i].cmp(&b[j]) {
                Ordering::Less => {
                    out.push(a[i].clone());
                    i += 1;
                }
                Ordering::Greater => {
                    out.push(b[j].clone());
                    j += 1;
                }
                Ordering::Equal => {
                    out.push(a[i].clone()); // duplicate → keep one
                    i += 1;
                    j += 1;
                }
            }
        }
        out.extend_from_slice(&a[i..]);
        out.extend_from_slice(&b[j..]);
        GSet { items: out }
    }

    /// Add one element (`union` with a singleton); idempotent if already present.
    #[must_use]
    pub fn add(&self, x: T) -> Self {
        if self.contains(&x) {
            self.clone()
        } else {
            self.union(&GSet::singleton(x))
        }
    }

    /// The elements in canonical (ascending) order — a defensive copy.
    #[must_use]
    pub fn to_vec(&self) -> Vec<T> {
        self.items.clone()
    }

    /// The canonical run as a slice (no copy).
    #[must_use]
    pub fn as_slice(&self) -> &[T] {
        &self.items
    }

    /// The number of elements.
    #[must_use]
    pub fn len(&self) -> usize {
        self.items.len()
    }

    /// Whether the set is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }
}

// ── generic-math additive-monoid surface (native Rust idiom, zero-dep) ──────
// G-Set is an additive, commutative + idempotent monoid (identity + associative
// `union`, NO inverse), so it surfaces only `Add` (the `+` operator) + a `Default`
// identity (the `Zero` analog) + `Sum` (fold a collection) — never `Sub`/`Neg`/`Mul`.
// `Default` is std's identity-value trait (no `num_traits` dep, per the zero-prod-dep
// doctrine); `+` IS `union`; `Sum` folds via the monoid so `iter.sum()` works.
//
// `Add` is implemented for `&GSet<T>` (ref-operator, `&a + &b`), NOT `GSet<T>` by
// value: the type already has an inherent `add(&self, x: T)` (add ONE element), and a
// by-value `Add for GSet` would let `Add::add(self, Self)` win method resolution over
// that inherent method for owned receivers — silently breaking `s.add(element)`. The
// ref-operator is the idiomatic Rust form for non-`Copy` collections and keeps both.

impl<T: Ord + Clone> Add for &GSet<T> {
    type Output = GSet<T>;

    /// The additive operator (`&a + &b`): the CRDT `union` merge — same merge as
    /// [`GSet::union`]. Ref-operator (see the module note) so it never collides with
    /// the inherent element-wise [`GSet::add`].
    fn add(self, rhs: &GSet<T>) -> GSet<T> {
        self.union(rhs)
    }
}

impl<T: Ord + Clone> Default for GSet<T> {
    /// The additive-monoid identity (the empty set) — the Rust `Default` analog of
    /// F# `Zero` / C# `AdditiveIdentity`.
    fn default() -> Self {
        Self::empty()
    }
}

impl<T: Ord + Clone> Sum for GSet<T> {
    /// Fold a collection of G-Sets through the monoid (`Default` identity + `+`), so
    /// `iter.sum()` aggregates them — the "generic code folds it for free" payoff.
    fn sum<I: Iterator<Item = GSet<T>>>(iter: I) -> Self {
        iter.fold(Self::default(), |acc, x| &acc + &x)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn g(xs: &[&str]) -> GSet<String> {
        GSet::of_iter(xs.iter().map(|s| (*s).to_string()))
    }

    #[test]
    fn of_iter_sorts_and_dedups() {
        assert_eq!(g(&["b", "a", "b", "c", "a"]).to_vec(), vec!["a", "b", "c"]);
    }

    #[test]
    fn contains_via_binary_search() {
        let s = g(&["a", "c", "e"]);
        assert!(s.contains(&"c".to_string()));
        assert!(!s.contains(&"d".to_string()));
    }

    #[test]
    fn add_is_idempotent() {
        let s = g(&["a", "b"]);
        assert_eq!(s.add("b".to_string()), s); // already present → no change
        assert_eq!(s.add("c".to_string()).to_vec(), vec!["a", "b", "c"]);
    }

    // The three CRDT convergence laws + identity (the fixture's `laws` block).
    #[test]
    fn law_idempotent() {
        let a = g(&["x", "y"]);
        assert_eq!(a.union(&a), a);
    }

    #[test]
    fn law_commutative() {
        let a = g(&["a", "c"]);
        let b = g(&["b", "c", "d"]);
        assert_eq!(a.union(&b), b.union(&a));
    }

    #[test]
    fn law_associative() {
        let a = g(&["a"]);
        let b = g(&["b", "a"]);
        let c = g(&["c", "b"]);
        assert_eq!(a.union(&b).union(&c), a.union(&b.union(&c)));
    }

    #[test]
    fn law_identity() {
        let a = g(&["a", "b"]);
        assert_eq!(a.union(&GSet::empty()), a);
        assert_eq!(GSet::<String>::empty().union(&a), a);
    }

    // ── generic-math additive-monoid surface (Add + Default + Sum) ──────────

    #[test]
    fn generic_math_add_merges_to_explicit_result() {
        // assert the explicit merge content (not `&a + &b == union`, which would be
        // tautological since `+` delegates to `union`) — proves `+` produces the merge
        let a = g(&["a", "b"]);
        let b = g(&["b", "c"]);
        assert_eq!((&a + &b).to_vec(), vec!["a", "b", "c"]);
    }

    #[test]
    fn generic_math_default_is_empty_identity() {
        let a = g(&["a", "b"]);
        let id = GSet::<String>::default();
        assert!(id.is_empty());
        assert_eq!(&id + &a, a); // identity + a = a
        assert_eq!(&a + &id, a); // a + identity = a
    }

    #[test]
    fn generic_math_monoid_laws_via_operator() {
        let a = g(&["a", "b"]);
        let b = g(&["b", "c"]);
        let c = g(&["c", "d"]);
        assert_eq!(&a + &a, a); // idempotent
        assert_eq!(&a + &b, &b + &a); // commutative
        assert_eq!(&(&a + &b) + &c, &a + &(&b + &c)); // associative
    }

    #[test]
    fn generic_math_sum_folds_via_monoid() {
        let sets = vec![g(&["a"]), g(&["b", "a"]), g(&["c"])];
        let merged: GSet<String> = sets.into_iter().sum();
        assert_eq!(merged.to_vec(), vec!["a", "b", "c"]);
    }
}

File: src/Core.CSharp/GSet.cs
// G-Set — grow-only set CRDT (bottom rung of the algebra ladder: G-Set ⊂ Bag ⊂ Z-set).
// C# parity oracle; mirrors src/Core/GSet.fs, src/Core.TypeScript/g-set, and
// src/Core.Rust.Algebra/src/gset.rs. Lives in the C# core next to the Z-set binding
// (ZetaCircuitBuilder), mirroring how F# keeps GSet.fs in src/Core/.

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Diagnostics.CodeAnalysis;
using System.Numerics;

namespace Zeta.Core.CSharp;

/// <summary>
/// Factory methods for <see cref="GSet{T}"/>. A non-generic companion class (the .NET
/// convention, e.g. <see cref="ImmutableArray"/> vs <see cref="ImmutableArray{T}"/>) so
/// the constructors stay static-member-free on the generic type (CA1000).
/// </summary>
public static class GSet
{
    /// <summary>The empty G-Set (the <see cref="GSet{T}.Union"/> identity) with a named collation.</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="collation">The named collation to use.</param>
    /// <returns>An empty set.</returns>
    public static GSet<T> Empty<T>(string collation) =>
        new(ImmutableArray<T>.Empty, Collation.ForKey<T>(collation), collation);

    /// <summary>The empty G-Set (the <see cref="GSet{T}.Union"/> identity).</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="comparer">Order on <typeparamref name="T"/>; defaults to default collation.</param>
    /// <returns>An empty set.</returns>
    public static GSet<T> Empty<T>(IComparer<T>? comparer = null) =>
        new(ImmutableArray<T>.Empty, comparer ?? Collation.ForKey<T>(), comparer == null ? Collation.DefaultName : "custom");

    /// <summary>The empty G-Set with a custom comparer and collation name.</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="comparer">Order on <typeparamref name="T"/>.</param>
    /// <param name="collation">The name of the collation.</param>
    /// <returns>An empty set.</returns>
    public static GSet<T> Empty<T>(IComparer<T> comparer, string collation) =>
        new(ImmutableArray<T>.Empty, comparer ?? Collation.ForKey<T>(), collation);

    /// <summary>A one-element G-Set with a named collation.</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="x">The single element.</param>
    /// <param name="collation">The named collation to use.</param>
    /// <returns>A set containing exactly <paramref name="x"/>.</returns>
    public static GSet<T> Singleton<T>(T x, string collation) =>
        new(ImmutableArray.Create(x), Collation.ForKey<T>(collation), collation);

    /// <summary>A one-element G-Set.</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="x">The single element.</param>
    /// <param name="comparer">Order on <typeparamref name="T"/>; defaults to default collation.</param>
    /// <returns>A set containing exactly <paramref name="x"/>.</returns>
    public static GSet<T> Singleton<T>(T x, IComparer<T>? comparer = null) =>
        new(ImmutableArray.Create(x), comparer ?? Collation.ForKey<T>(), comparer == null ? Collation.DefaultName : "custom");

    /// <summary>A one-element G-Set with a custom comparer and collation name.</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="x">The single element.</param>
    /// <param name="comparer">Order on <typeparamref name="T"/>.</param>
    /// <param name="collation">The name of the collation.</param>
    /// <returns>A set containing exactly <paramref name="x"/>.</returns>
    public static GSet<T> Singleton<T>(T x, IComparer<T> comparer, string collation) =>
        new(ImmutableArray.Create(x), comparer ?? Collation.ForKey<T>(), collation);

    /// <summary>Canonicalize arbitrary input with a named collation.</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="xs">The elements.</param>
    /// <param name="collation">The named collation to use.</param>
    /// <returns>The canonical G-Set of the distinct elements.</returns>
    public static GSet<T> OfSeq<T>(IEnumerable<T> xs, string collation)
    {
        var cmp = Collation.ForKey<T>(collation);
        return new GSet<T>(Canonicalize(xs, cmp), cmp, collation);
    }

    /// <summary>Canonicalize arbitrary input.</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="xs">The elements.</param>
    /// <param name="comparer">Order on <typeparamref name="T"/>; defaults to default collation.</param>
    /// <returns>The canonical G-Set of the distinct elements.</returns>
    public static GSet<T> OfSeq<T>(IEnumerable<T> xs, IComparer<T>? comparer = null)
    {
        var cmp = comparer ?? Collation.ForKey<T>();
        return new GSet<T>(Canonicalize(xs, cmp), cmp, comparer == null ? Collation.DefaultName : "custom");
    }

    /// <summary>Canonicalize arbitrary input with a custom comparer and collation name.</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="xs">The elements.</param>
    /// <param name="comparer">Order on <typeparamref name="T"/>.</param>
    /// <param name="collation">The name of the collation.</param>
    /// <returns>The canonical G-Set of the distinct elements.</returns>
    public static GSet<T> OfSeq<T>(IEnumerable<T> xs, IComparer<T> comparer, string collation)
    {
        var cmp = comparer ?? Collation.ForKey<T>();
        return new GSet<T>(Canonicalize(xs, cmp), cmp, collation);
    }

    /// <summary>Sort ascending + drop duplicates under <paramref name="cmp"/> (the canonical run).</summary>
    /// <typeparam name="T">The element type.</typeparam>
    /// <param name="xs">The elements.</param>
    /// <param name="cmp">The order.</param>
    /// <returns>The canonical ascending, duplicate-free run.</returns>
    internal static ImmutableArray<T> Canonicalize<T>(IEnumerable<T> xs, IComparer<T> cmp)
    {
        var sorted = new List<T>(xs);
        sorted.Sort(cmp);
        var builder = ImmutableArray.CreateBuilder<T>(sorted.Count);
        for (var k = 0; k < sorted.Count; k++)
        {
            if (k == 0 || cmp.Compare(sorted[k - 1], sorted[k]) != 0)
            {
                builder.Add(sorted[k]);
            }
        }

        return builder.ToImmutable();
    }
}

/// <summary>
/// A grow-only set (G-Set CRDT): a canonical ascending-sorted, duplicate-free run under
/// an <see cref="IComparer{T}"/>. <see cref="Union"/> is the only combiner and is
/// idempotent, commutative, and associative (the three CRDT convergence laws), so
/// replicas that have seen the same elements — in any order — converge with no
/// coordination. Parity with <c>src/Core/GSet.fs</c>, <c>src/Core.TypeScript/g-set</c>,
/// and <c>src/Core.Rust.Algebra/src/gset.rs</c>. Construct via <see cref="GSet"/>.
/// </summary>
/// <remarks>
/// The comparer is explicit (like the TS <c>compare</c> parameter) rather than baked to
/// <see cref="Comparer{T}.Default"/>: for <see cref="string"/> the default is
/// culture-sensitive, but the cross-language wire needs the project binary collation
/// (Unicode code-point / UTF-8 byte order). Pass it explicitly for cross-language parity.
/// </remarks>
/// <typeparam name="T">The element type.</typeparam>
public sealed class GSet<T> :
    IEquatable<GSet<T>>,
    IAdditiveIdentity<GSet<T>, GSet<T>>,
    IAdditionOperators<GSet<T>, GSet<T>, GSet<T>>
{
    private readonly ImmutableArray<T> _items;
    private readonly IComparer<T> _comparer;

    /// <summary>Gets the name of the collation used by this set.</summary>
    public string CollationName { get; }

    internal GSet(ImmutableArray<T> items, IComparer<T> comparer, string collationName)
    {
        _items = items.IsDefault ? ImmutableArray<T>.Empty : items;
        _comparer = comparer;
        CollationName = collationName;
    }

    /// <summary>The number of elements.</summary>
    public int Count => _items.Length;

    /// <summary>Whether the set is empty.</summary>
    public bool IsEmpty => _items.IsEmpty;

    /// <summary>Membership via binary search on the sorted run. O(log n).</summary>
    /// <param name="x">The element to test.</param>
    /// <returns><see langword="true"/> iff <paramref name="x"/> is present.</returns>
    public bool Contains(T x)
    {
        int lo = 0, hi = _items.Length - 1;
        while (lo <= hi)
        {
            var mid = lo + ((hi - lo) >> 1);
            var c = _comparer.Compare(_items[mid], x);
            if (c == 0)
            {
                return true;
            }

            if (c < 0)
            {
                lo = mid + 1;
            }
            else
            {
                hi = mid - 1;
            }
        }

        return false;
    }

    /// <summary>
    /// The CRDT merge: the union of two sorted-unique runs, kept sorted-unique.
    /// Idempotent, commutative, associative. Uses this set's comparer.
    /// </summary>
    /// <param name="other">The set to merge with.</param>
    /// <returns>The union, in canonical order.</returns>
    public GSet<T> Union(GSet<T> other)
    {
        ArgumentNullException.ThrowIfNull(other);

        // The comparer is part of a set's identity: two sets with different comparers are
        // not union-compatible. Silently recanonicalizing would break commutativity
        // (a∪b under a's order ≠ b∪a under b's order), so fail fast on mismatch
        // (PR review 2026-06-01). Same comparer ⇒ both runs are sorted alike ⇒ linear merge.
        RequireSameComparer(other);

        if (_items.IsEmpty)
        {
            return other;
        }

        if (other._items.IsEmpty)
        {
            return this;
        }

        var a = _items;
        var b = other._items;
        var builder = ImmutableArray.CreateBuilder<T>(a.Length + b.Length);
        int i = 0, j = 0;
        while (i < a.Length && j < b.Length)
        {
            var c = _comparer.Compare(a[i], b[j]);
            if (c < 0)
            {
                builder.Add(a[i]);
                i++;
            }
            else if (c > 0)
            {
                builder.Add(b[j]);
                j++;
            }
            else
            {
                builder.Add(a[i]); // duplicate → keep one
                i++;
                j++;
            }
        }

        while (i < a.Length)
        {
            builder.Add(a[i]);
            i++;
        }

        while (j < b.Length)
        {
            builder.Add(b[j]);
            j++;
        }

        return new GSet<T>(builder.ToImmutable(), _comparer, CollationName);
    }

    /// <summary>
    /// The additive-monoid identity (generic-math <see cref="IAdditiveIdentity{TSelf, TResult}"/>):
    /// the empty set (default comparer). G-Set is an additive, commutative + idempotent monoid
    /// (identity + associative <see cref="Union"/>, NO inverse), so it surfaces only
    /// <see cref="IAdditiveIdentity{TSelf, TResult}"/> + <see cref="IAdditionOperators{TSelf, TOther, TResult}"/>
    /// — never <c>INumber</c> (no negation / order / multiplication).
    /// </summary>
    [SuppressMessage(
        "Design",
        "CA1000:Do not declare static members on generic types",
        Justification = "IAdditiveIdentity<TSelf,TResult> requires a static AdditiveIdentity member on the generic type; CA1000 predates static-abstract interface members (IWSAM) and does not apply to generic-math interface implementations.")]
    public static GSet<T> AdditiveIdentity { get; } = GSet.Empty<T>();

    /// <summary>
    /// The additive operator (generic-math <c>+</c>): the CRDT <see cref="Union"/> merge.
    /// The empty set acts as a comparer-agnostic identity (<c>empty + x = x</c>, <c>x + empty = x</c>)
    /// so <see cref="AdditiveIdentity"/> composes with a set under any comparer; two NON-empty
    /// operands still delegate to <see cref="Union"/>, which enforces the comparer-identity match.
    /// </summary>
    /// <param name="left">The left set.</param>
    /// <param name="right">The right set.</param>
    /// <returns>The union, in canonical order.</returns>
    public static GSet<T> operator +(GSet<T> left, GSet<T> right)
    {
        ArgumentNullException.ThrowIfNull(left);
        ArgumentNullException.ThrowIfNull(right);

        // The additive identity is the empty set; it must absorb under ANY comparer for the
        // monoid identity law to hold (AdditiveIdentity uses the default comparer, but a set
        // may carry a custom one). Empty has no elements, so its ordering is irrelevant —
        // short-circuit before Union's deliberate same-comparer check (guards non-empty merges).
        if (left.IsEmpty)
        {
            return right;
        }

        if (right.IsEmpty)
        {
            return left;
        }

        return left.Union(right);
    }

    /// <summary>Add one element (<see cref="Union"/> with a singleton); idempotent if already present.</summary>
    /// <param name="x">The element to add.</param>
    /// <returns>The set with <paramref name="x"/> present.</returns>
    public GSet<T> Add(T x) =>
        Contains(x) ? this : Union(new GSet<T>(ImmutableArray.Create(x), _comparer, CollationName));

    private void RequireSameComparer(GSet<T> other)
    {
        var sameCollation = !string.Equals(CollationName, "custom", StringComparison.Ordinal) &&
                            string.Equals(CollationName, other.CollationName, StringComparison.Ordinal);

        if (!sameCollation && !_comparer.Equals(other._comparer))
        {
            throw new ArgumentException(
                $"GSet.Union requires both sets to use the same collation name or equivalent comparer (this: '{CollationName}', other: '{other.CollationName}').",
                nameof(other));
        }
    }

    /// <summary>The canonical (ascending) elements as an immutable array.</summary>
    /// <returns>The canonical run.</returns>
    public ImmutableArray<T> ToImmutableArray() => _items;

    /// <summary>The canonical (ascending) elements as a new array.</summary>
    /// <returns>A fresh array of the elements in canonical order.</returns>
    public T[] ToArray() => _items.AsSpan().ToArray();

    /// <summary>
    /// Value equality: same elements in canonical order (element-wise under the
    /// comparer). Because both sets are canonical, this is set equality.
    /// </summary>
    /// <param name="other">The set to compare.</param>
    /// <returns><see langword="true"/> iff the sets hold the same elements.</returns>
    public bool Equals(GSet<T>? other)
    {
        if (other is null)
        {
            return false;
        }

        if (ReferenceEquals(this, other))
        {
            return true;
        }

        var sameCollation = !string.Equals(CollationName, "custom", StringComparison.Ordinal) &&
                            string.Equals(CollationName, other.CollationName, StringComparison.Ordinal);

        if (!sameCollation && !_comparer.Equals(other._comparer))
        {
            return false;
        }

        if (_items.Length != other._items.Length)
        {
            return false;
        }

        for (var k = 0; k < _items.Length; k++)
        {
            if (_comparer.Compare(_items[k], other._items[k]) != 0)
            {
                return false;
            }
        }

        return true;
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => Equals(obj as GSet<T>);

    /// <inheritdoc/>
    public override int GetHashCode()
    {
        // Consistent with Equals (which includes the comparer in identity): hash the
        // comparer + the elements. When the comparer is also an equality comparer (e.g.
        // StringComparer.Ordinal) hash each element through it so Compare==0 values hash
        // alike; otherwise count-only for the element part (still consistent with the
        // comparer-based Equals). (PR review 2026-06-01.)
        var hash = default(HashCode);
        hash.Add(CollationName, StringComparer.Ordinal);
        hash.Add(_comparer);
        hash.Add(_items.Length);
        if (_comparer is IEqualityComparer<T> eq)
        {
            foreach (var x in _items)
            {
                hash.Add(x, eq);
            }
        }

        return hash.ToHashCode();
    }
}
